torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,438 @@
+  // as the rightmost.
+  std::vector<Val*> sortedInputs(const std::list<ValInfo>& variables) {
+    std::unordered_set<Val*> variables_set;
+    variables_set.reserve(variables.size());
+    for (auto v : variables) {
+      variables_set.emplace(v.variable);
+    }
+    std::vector<Val*> sorted_inputs(inputs().begin(), inputs().end());
+    std::unordered_map<Val*, std::unordered_set<Val*>> dependency;
+    dependency.reserve(sorted_inputs.size());
+    for (auto v : sorted_inputs) {
+      dependency[v] = getSubexprDependency(v, variables_set);
+    }
+    auto compare = [&](Val* v1, Val* v2) {
+      // Find all variables in variables_set that v1 and v2 depends on. The
+      // input (v1 or v2) that exclusively has the right most variable in
+      // variables_set will be to the right of the other input.
+      bool v1_is_left_of_v2 = false;
+      auto deps1 = dependency.at(v1);
+      auto deps2 = dependency.at(v2);
+      if (hasTensor(deps2)) {
+        return true;
+      }
+      if (hasTensor(deps1)) {
+        return false;
+      }
+      for (auto v : variables) {
+        if (deps1.count(v.variable) > 0 && deps2.count(v.variable) == 0) {
+          v1_is_left_of_v2 = false;
+        } else if (
+            deps2.count(v.variable) > 0 && deps1.count(v.variable) == 0) {
+          v1_is_left_of_v2 = true;
+        }
+      }
+      return v1_is_left_of_v2;
+    };
+    std::sort(sorted_inputs.begin(), sorted_inputs.end(), compare);
+    return sorted_inputs;
+  }
+
+ protected:
+  // Add a new value as an input of this expression. If `value` is a constant
+  // scalar, then evalate this constant and merge it with the constant term.
+  // Otherwise, add it as an input.
+  void addInput(Val* value) {
+    TORCH_CHECK(
+        hasSimilarType(dtype(), *value->getDataType()),
+        "Input types should be similar, but got: ",
+        dtype(),
+        ", and ",
+        *value->getDataType());
+    value = foldConstants(value);
+    if (value->isConst()) {
+      updateConstantTermWith(value);
+    } else {
+      Expr::addInput(value);
+    }
+  }
+
+  // Update the constant term. For example, if getOpType() == "+", and the
+  // current constant term is 8, then updateConstantTermWith(7) will change the
+  // constant term to 15.
+  void updateConstantTermWith(Val* value) {
+    if (!hasConstantTerm()) {
+      addAttribute(value);
+      return;
+    }
+    auto new_constant_term = IrBuilder::newScalar(dtype());
+    IrBuilder::create<BinaryOp>(
+        getOpType(), new_constant_term, getConstantTerm(), value);
+    new_constant_term = foldConstants(new_constant_term);
+    TORCH_INTERNAL_ASSERT(new_constant_term->isConst());
+    attributes_.at(1) = new_constant_term;
+  }
+};
+
+NVFUSER_DEFINE_CLONE_AND_CREATE(FlattenedAssocCommOp)
+
+// Recursively convert expressions like AddOp(AddOp(a, b), AddOp(c, d)) into
+// FlattenedAdd(a, b, c, d). This function recursives transforms the entire
```

Comment:
typo: recursively

Reply:
fixed in https://github.com/csarofeen/pytorch/pull/2257